src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,157 @@
+
+                foreach (var fixForSpan in fixesForSpan)
+                {
+                    // If there is more than one fix then we will leave it for the user to manually apply
+                    if (fixForSpan.Fixes.Length == 1)
+                    {
+                        fixes.Add(fixForSpan.Fixes[0]);
+                    }
+                }
+            }
+
+            return fixes.ToImmutableAndFree();
+        }
+
+        private async Task<Document> ApplyFixesAsync(Document document, ImmutableArray<AddImportFixData> fixes, CancellationToken cancellationToken)
+        {
+            if (fixes.IsEmpty
+                || !document.TryGetText(out var text))
+            {
+                return document;
+            }
+
+            var solution = document.Project.Solution;
+            var packageInstallerService = solution.Workspace.Services.GetService<IPackageInstallerService>();
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            // Do not limit the results since we plan to fix all the reported issues.
+            var codeActions = addImportService.GetCodeActionsForFixes(document, fixes, packageInstallerService, limitResults: false);
+            var getChangesTasks = codeActions.Select(action => GetChangesForCodeActionAsync(document, action, cancellationToken));
+
+            await Task.WhenAll(getChangesTasks).ConfigureAwait(false);
+
+            var allTextChanges = Enumerable.Empty<TextChange>();
+            var allAddedProjectReferences = Enumerable.Empty<ProjectReference>();
+            var allAddedMetaDataReferences = Enumerable.Empty<MetadataReference>();
+
+            foreach (var getChangesTask in getChangesTasks)
+            {
+                var (projectChanges, textChanges) = await getChangesTask.ConfigureAwait(false);
+
+                allTextChanges = allTextChanges.Concat(textChanges);
+                allAddedProjectReferences = allAddedProjectReferences.Concat(projectChanges.GetAddedProjectReferences());
+                allAddedMetaDataReferences = allAddedMetaDataReferences.Concat(projectChanges.GetAddedMetadataReferences());
+            }
+
+            var newProject = document.Project.AddMetadataReferences(allAddedMetaDataReferences.Distinct());
+            newProject = newProject.AddProjectReferences(allAddedProjectReferences.Distinct());
+
+            var newText = text.WithChanges(allTextChanges.Distinct());
+            var newDocument = newProject.GetDocument(document.Id).WithText(newText);
+
+            return newDocument;
+        }
+
+        private async Task<(ProjectChanges projectChanges, IEnumerable<TextChange> textChanges)> GetChangesForCodeActionAsync(Document document, CodeAction codeAction, CancellationToken cancellationToken)
+        {
+            var newSolution = await codeAction.GetChangedSolutionAsync(new ProgressTracker(), cancellationToken).ConfigureAwait(false);
```

Comment:
So the only code using this service at the moment is the `AddMissingImportsRefactoringProvider`. The `CodeRefactoringContext` that is passed in does not have a progress tracker. Meaning none of the calling code has a progress tracker. Should I instead use the `GetChangedSolutionInternalAsync` method since this is acting as a batch fixer of sorts?